# Exploratory Data Analysis

In [72]:
import pandas as pd
import numpy as np
import seaborn as sns
import scipy as scipy
# import matplotlib.pylib as plt
import matplotlib.pyplot as plt
import sqlite3
plt.style.use('ggplot')

## Load Data

In [73]:
cnx = sqlite3.connect('data/agri.db') # Establish connection to SQLite database file agri.db

df = pd.read_sql_query("SELECT * FROM farm_data", cnx) # Retrieve all rows and columns from farm_data table in the database and load result into Pandas dataframe df
df # Dataframe

System Location Code Previous Cycle Plant Type           Plant Type  \
0                   Zone_D                     Herbs           Vine Crops   
1                   Zone_G                     Herbs           VINE CROPS   
2                   Zone_F                     Herbs                herbs   
3                   Zone_G                     Herbs  fruiting vegetables   
4                   Zone_B                Vine Crops         LEAFY GREENS   
...                    ...                       ...                  ...   
57484               Zone_C                Vine Crops  Fruiting Vegetables   
57485               Zone_C                Vine Crops  Fruiting Vegetables   
57486               Zone_E                     Herbs           VINE CROPS   
57487               Zone_G       Fruiting Vegetables           VINE CROPS   
57488               Zone_F       Fruiting Vegetables         Leafy Greens   

      Plant Stage  Temperature Sensor (°C)  Humidity Sensor (%)  \
0        Maturity                    23.34                  NaN   
1      Vegetative                   -24.16                63.14   
2        maturity                    23.84                  NaN   
3      Vegetative                    22.61                  NaN   
4        seedling                    22.88                  NaN   
...           ...                      ...                  ...   
57484    Maturity                      NaN                  NaN   
57485    Maturity                      NaN                  NaN   
57486    Seedling                    24.85                  NaN   
57487  Vegetative                    23.45                64.89   
57488    Maturity                    20.00                  NaN   

       Light Intensity Sensor (lux)  CO2 Sensor (ppm)  EC Sensor (dS/m)  \
0                             454.0               937              1.82   
1                             441.0               938              2.13   
2                             562.0              1246              1.21   
3                             651.0              1346              1.96   
4                             144.0               812              2.76   
...                             ...               ...               ...   
57484                         619.0              1074              1.95   
57485                         619.0              1074              1.95   
57486                         203.0               816              0.98   
57487                         502.0               911              2.27   
57488                         391.0               864              1.90   

       O2 Sensor (ppm) Nutrient N Sensor (ppm) Nutrient P Sensor (ppm)  \
0                    4                     161                      64   
1                    6                     161                      53   
2                    8                     228                      79   
3                    6                    None                  53 ppm   
4                    5                      61                      19   
...                ...                     ...                     ...   
57484                6                     150                      61   
57485                6                     150                      61   
57486                7                     121                      41   
57487                5                     190                      57   
57488                7                     108                      40   

      Nutrient K Sensor (ppm)  pH Sensor  Water Level Sensor (mm)  
0                         201        6.1                28.150346  
1                        None        6.5                31.000000  
2                         292        6.4                26.000000  
3                         289        6.6                      NaN  
4                         168        5.5                28.000000  
...                       ...        ...                      ...  
57484     

## Data Understanding

In [ ]:
df.shape # Understand size of dataset

(57489, 15)

We see that the dataframe has 57489 rows and 15 columns.

In [75]:
df.columns # List out columns for easy reference

Index(['System Location Code', 'Previous Cycle Plant Type', 'Plant Type',
       'Plant Stage', 'Temperature Sensor (°C)', 'Humidity Sensor (%)',
       'Light Intensity Sensor (lux)', 'CO2 Sensor (ppm)', 'EC Sensor (dS/m)',
       'O2 Sensor (ppm)', 'Nutrient N Sensor (ppm)', 'Nutrient P Sensor (ppm)',
       'Nutrient K Sensor (ppm)', 'pH Sensor', 'Water Level Sensor (mm)'],
      dtype='object')

In [76]:
df.info() # List the column names and their respective data types, show if there are Null entries in each column to spot improperly formatted and missing data

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 57489 entries, 0 to 57488
Data columns (total 15 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   System Location Code          57489 non-null  object 
 1   Previous Cycle Plant Type     57489 non-null  object 
 2   Plant Type                    57489 non-null  object 
 3   Plant Stage                   57489 non-null  object 
 4   Temperature Sensor (°C)       48800 non-null  float64
 5   Humidity Sensor (%)           18622 non-null  float64
 6   Light Intensity Sensor (lux)  53211 non-null  float64
 7   CO2 Sensor (ppm)              57489 non-null  int64  
 8   EC Sensor (dS/m)              57489 non-null  float64
 9   O2 Sensor (ppm)               57489 non-null  int64  
 10  Nutrient N Sensor (ppm)       47515 non-null  object 
 11  Nutrient P Sensor (ppm)       51791 non-null  object 
 12  Nutrient K Sensor (ppm)       53788 non-null  object 
 13  p

Only the first four columns 'System Location Code', 'Previous Cycle Plant Type', 'Plant Type' and 'Plant Stage' have no Null values. The rest of the columns have multiple Null vaues that we will have to replace during data cleaning later.

## Data Cleaning

In [77]:
# Some entries in the Nutrient Sensor columns have "ppm" following the integers - let's strip that
df['Nutrient N Sensor (ppm)'] = df['Nutrient N Sensor (ppm)'].str.rstrip(" ppm")
df['Nutrient P Sensor (ppm)'] = df['Nutrient P Sensor (ppm)'].str.rstrip(" ppm")
df['Nutrient K Sensor (ppm)'] = df['Nutrient K Sensor (ppm)'].str.rstrip(" ppm")
df

System Location Code Previous Cycle Plant Type           Plant Type  \
0                   Zone_D                     Herbs           Vine Crops   
1                   Zone_G                     Herbs           VINE CROPS   
2                   Zone_F                     Herbs                herbs   
3                   Zone_G                     Herbs  fruiting vegetables   
4                   Zone_B                Vine Crops         LEAFY GREENS   
...                    ...                       ...                  ...   
57484               Zone_C                Vine Crops  Fruiting Vegetables   
57485               Zone_C                Vine Crops  Fruiting Vegetables   
57486               Zone_E                     Herbs           VINE CROPS   
57487               Zone_G       Fruiting Vegetables           VINE CROPS   
57488               Zone_F       Fruiting Vegetables         Leafy Greens   

      Plant Stage  Temperature Sensor (°C)  Humidity Sensor (%)  \
0        Maturity                    23.34                  NaN   
1      Vegetative                   -24.16                63.14   
2        maturity                    23.84                  NaN   
3      Vegetative                    22.61                  NaN   
4        seedling                    22.88                  NaN   
...           ...                      ...                  ...   
57484    Maturity                      NaN                  NaN   
57485    Maturity                      NaN                  NaN   
57486    Seedling                    24.85                  NaN   
57487  Vegetative                    23.45                64.89   
57488    Maturity                    20.00                  NaN   

       Light Intensity Sensor (lux)  CO2 Sensor (ppm)  EC Sensor (dS/m)  \
0                             454.0               937              1.82   
1                             441.0               938              2.13   
2                             562.0              1246              1.21   
3                             651.0              1346              1.96   
4                             144.0               812              2.76   
...                             ...               ...               ...   
57484                         619.0              1074              1.95   
57485                         619.0              1074              1.95   
57486                         203.0               816              0.98   
57487                         502.0               911              2.27   
57488                         391.0               864              1.90   

       O2 Sensor (ppm) Nutrient N Sensor (ppm) Nutrient P Sensor (ppm)  \
0                    4                     161                      64   
1                    6                     161                      53   
2                    8                     228                      79   
3                    6                    None                      53   
4                    5                      61                      19   
...                ...                     ...                     ...   
57484                6                     150                      61   
57485                6                     150                      61   
57486                7                     121                      41   
57487                5                     190                      57   
57488                7                     108                      40   

      Nutrient K Sensor (ppm)  pH Sensor  Water Level Sensor (mm)  
0                         201        6.1                28.150346  
1                        None        6.5                31.000000  
2                         292        6.4                26.000000  
3                         289        6.6                      NaN  
4                         168        5.5                28.000000  
...                       ...        ...                      ...  
57484     

In [78]:
df["System Location Code"].value_counts() # Check unique values to identify any inconsistencies or anomalies in System Location Code - all good here

System Location Code
Zone_B    8296
Zone_G    8287
Zone_E    8272
Zone_C    8268
Zone_F    8156
Zone_A    8148
Zone_D    8062
Name: count, dtype: int64

In [79]:
df["Previous Cycle Plant Type"].value_counts() # Check unique values to identify any inconsistencies or anomalies in Previous Cycle Plant Type - all good here

Previous Cycle Plant Type
Fruiting Vegetables    14481
Leafy Greens           14446
Vine Crops             14305
Herbs                  14257
Name: count, dtype: int64

In [80]:
df["Plant Type"].value_counts() # Check unique values to identify any inconsistencies or anomalies in Plant Type - we can see different capitalization styles for similar categories of Plant Type


Plant Type
Fruiting Vegetables    10137
Herbs                  10059
Leafy Greens            9968
Vine Crops              9921
leafy greens            2241
HERBS                   2239
LEAFY GREENS            2183
VINE CROPS              2173
vine crops              2163
FRUITING VEGETABLES     2154
fruiting vegetables     2128
herbs                   2123
Name: count, dtype: int64

In [87]:
df["Plant Type"] = df["Plant Type"].map(str.title) # Standardize the Plant Type column by converting each value to title case
df["Plant Type"].value_counts() # Check unique values - all good now

Plant Type
Herbs                  14421
Fruiting Vegetables    14419
Leafy Greens           14392
Vine Crops             14257
Name: count, dtype: int64

In [88]:
df["Plant Stage"].value_counts() # Check unique values to identify any inconsistencies or anomalies in Plant Stage - we can see different capitalization styles for similar categories of Plant Stage

Plant Stage
Seedling      19425
Vegetative    19062
Maturity      19002
Name: count, dtype: int64

In [89]:
df["Plant Stage"] = df["Plant Stage"].map(str.title) # Standardize the Plant Stage column by converting each value to title case
df["Plant Stage"].value_counts() # Check unique values - all good now

Plant Stage
Seedling      19425
Vegetative    19062
Maturity      19002
Name: count, dtype: int64

In [84]:
df # Look at data frame after cleaning

System Location Code Previous Cycle Plant Type           Plant Type  \
0                   Zone_D                     Herbs           Vine Crops   
1                   Zone_G                     Herbs           Vine Crops   
2                   Zone_F                     Herbs                Herbs   
3                   Zone_G                     Herbs  Fruiting Vegetables   
4                   Zone_B                Vine Crops         Leafy Greens   
...                    ...                       ...                  ...   
57484               Zone_C                Vine Crops  Fruiting Vegetables   
57485               Zone_C                Vine Crops  Fruiting Vegetables   
57486               Zone_E                     Herbs           Vine Crops   
57487               Zone_G       Fruiting Vegetables           Vine Crops   
57488               Zone_F       Fruiting Vegetables         Leafy Greens   

      Plant Stage  Temperature Sensor (°C)  Humidity Sensor (%)  \
0        Maturity                    23.34                  NaN   
1      Vegetative                   -24.16                63.14   
2        Maturity                    23.84                  NaN   
3      Vegetative                    22.61                  NaN   
4        Seedling                    22.88                  NaN   
...           ...                      ...                  ...   
57484    Maturity                      NaN                  NaN   
57485    Maturity                      NaN                  NaN   
57486    Seedling                    24.85                  NaN   
57487  Vegetative                    23.45                64.89   
57488    Maturity                    20.00                  NaN   

       Light Intensity Sensor (lux)  CO2 Sensor (ppm)  EC Sensor (dS/m)  \
0                             454.0               937              1.82   
1                             441.0               938              2.13   
2                             562.0              1246              1.21   
3                             651.0              1346              1.96   
4                             144.0               812              2.76   
...                             ...               ...               ...   
57484                         619.0              1074              1.95   
57485                         619.0              1074              1.95   
57486                         203.0               816              0.98   
57487                         502.0               911              2.27   
57488                         391.0               864              1.90   

       O2 Sensor (ppm) Nutrient N Sensor (ppm) Nutrient P Sensor (ppm)  \
0                    4                     161                      64   
1                    6                     161                      53   
2                    8                     228                      79   
3                    6                    None                      53   
4                    5                      61                      19   
...                ...                     ...                     ...   
57484                6                     150                      61   
57485                6                     150                      61   
57486                7                     121                      41   
57487                5                     190                      57   
57488                7                     108                      40   

      Nutrient K Sensor (ppm)  pH Sensor  Water Level Sensor (mm)  
0                         201        6.1                28.150346  
1                        None        6.5                31.000000  
2                         292        6.4                26.000000  
3                         289        6.6                      NaN  
4                         168        5.5                28.000000  
...                       ...        ...                      ...  
57484     

In [85]:
df = df.fillna('') # Replace all NaN (missing) values in the dataframe with empty strings ('') to handle missing data
df

System Location Code Previous Cycle Plant Type           Plant Type  \
0                   Zone_D                     Herbs           Vine Crops   
1                   Zone_G                     Herbs           Vine Crops   
2                   Zone_F                     Herbs                Herbs   
3                   Zone_G                     Herbs  Fruiting Vegetables   
4                   Zone_B                Vine Crops         Leafy Greens   
...                    ...                       ...                  ...   
57484               Zone_C                Vine Crops  Fruiting Vegetables   
57485               Zone_C                Vine Crops  Fruiting Vegetables   
57486               Zone_E                     Herbs           Vine Crops   
57487               Zone_G       Fruiting Vegetables           Vine Crops   
57488               Zone_F       Fruiting Vegetables         Leafy Greens   

      Plant Stage Temperature Sensor (°C) Humidity Sensor (%)  \
0        Maturity                   23.34                       
1      Vegetative                  -24.16               63.14   
2        Maturity                   23.84                       
3      Vegetative                   22.61                       
4        Seedling                   22.88                       
...           ...                     ...                 ...   
57484    Maturity                                               
57485    Maturity                                               
57486    Seedling                   24.85                       
57487  Vegetative                   23.45               64.89   
57488    Maturity                    20.0                       

      Light Intensity Sensor (lux)  CO2 Sensor (ppm)  EC Sensor (dS/m)  \
0                            454.0               937              1.82   
1                            441.0               938              2.13   
2                            562.0              1246              1.21   
3                            651.0              1346              1.96   
4                            144.0               812              2.76   
...                            ...               ...               ...   
57484                        619.0              1074              1.95   
57485                        619.0              1074              1.95   
57486                        203.0               816              0.98   
57487                        502.0               911              2.27   
57488                        391.0               864              1.90   

       O2 Sensor (ppm) Nutrient N Sensor (ppm) Nutrient P Sensor (ppm)  \
0                    4                     161                      64   
1                    6                     161                      53   
2                    8                     228                      79   
3                    6                                              53   
4                    5                      61                      19   
...                ...                     ...                     ...   
57484                6                     150                      61   
57485                6                     150                      61   
57486                7                     121                      41   
57487                5                     190                      57   
57488                7                     108                      40   

      Nutrient K Sensor (ppm)  pH Sensor Water Level Sensor (mm)  
0                         201        6.1               28.150346  
1                                    6.5                    31.0  
2                         292        6.4                    26.0  
3                         289        6.6                          
4                         168        5.5                    28.0  
...                       ...        ...                     ...  
57484                     210        6.4             

In [92]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 57489 entries, 0 to 57488
Data columns (total 15 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   System Location Code          57489 non-null  object 
 1   Previous Cycle Plant Type     57489 non-null  object 
 2   Plant Type                    57489 non-null  object 
 3   Plant Stage                   57489 non-null  object 
 4   Temperature Sensor (°C)       57489 non-null  object 
 5   Humidity Sensor (%)           57489 non-null  object 
 6   Light Intensity Sensor (lux)  57489 non-null  object 
 7   CO2 Sensor (ppm)              57489 non-null  int64  
 8   EC Sensor (dS/m)              57489 non-null  float64
 9   O2 Sensor (ppm)               57489 non-null  int64  
 10  Nutrient N Sensor (ppm)       57489 non-null  object 
 11  Nutrient P Sensor (ppm)       57489 non-null  object 
 12  Nutrient K Sensor (ppm)       57489 non-null  object 
 13  p

In [ ]:
df.isna().sum() # check if there are any more NaN (missing) values

System Location Code            0
Previous Cycle Plant Type       0
Plant Type                      0
Plant Stage                     0
Temperature Sensor (°C)         0
Humidity Sensor (%)             0
Light Intensity Sensor (lux)    0
CO2 Sensor (ppm)                0
EC Sensor (dS/m)                0
O2 Sensor (ppm)                 0
Nutrient N Sensor (ppm)         0
Nutrient P Sensor (ppm)         0
Nutrient K Sensor (ppm)         0
pH Sensor                       0
Water Level Sensor (mm)         0
dtype: int64

In [86]:
df.describe()

CO2 Sensor (ppm)  EC Sensor (dS/m)  O2 Sensor (ppm)     pH Sensor
count      57489.000000      57489.000000     57489.000000  57489.000000
mean        1083.376576          1.624435         6.580842      6.021446
std          172.570351          0.415535         1.145534      0.400251
min          799.000000         -0.250000         3.000000      3.800000
25%          950.000000          1.360000         6.000000      5.800000
50%         1071.000000          1.650000         7.000000      6.000000
75%         1178.000000          1.900000         7.000000      6.300000
max         1500.000000          3.440000        11.000000      7.800000